universal-application-tool-0.0.1/app/controllers/LoginController.java

```diff
@@ -33,88 +30,70 @@
 
   private final Config config;
 
-  private final String applicantIdp;
-
   @Inject
   public LoginController(
-      @AdOidcClient @Nullable OidcClient adClient,
-      @IdcsOidcClient @Nullable OidcClient idcsClient,
-      @LoginRadiusSamlClient @Nullable SAML2Client loginRadiusClient,
+      @AdminAuthClient @Nullable IndirectClient adminClient,
+      @ApplicantAuthClient @Nullable IndirectClient applicantClient,
       SessionStore sessionStore,
       Config config) {
-    this.idcsClient = idcsClient;
-    this.adClient = adClient;
-    this.loginRadiusClient = loginRadiusClient;
+    this.adminClient = adminClient;
+    this.applicantClient = applicantClient;
     this.sessionStore = Preconditions.checkNotNull(sessionStore);
-    this.applicantIdp = checkNotNull(config).getString("auth.applicant_idp");
     this.httpActionAdapter = PlayHttpActionAdapter.INSTANCE;
     this.config = config;
   }
 
-  public Result loginWithRedirect(Http.Request request, Optional<String> redirectTo) {
-    if (applicantIdp.equals(AuthIdentityProviderName.LOGIN_RADIUS_APPLICANT.getString())) {
-      return loginRadiusLoginWithRedirect(request, redirectTo);
-    }
-    return idcsLoginWithRedirect(request, redirectTo);
-  }
-
-  public Result idcsLogin(Http.Request request) {
-    return login(request, idcsClient);
-  }
-
-  public Result idcsLoginWithRedirect(Http.Request request, Optional<String> redirectTo) {
-    if (redirectTo.isEmpty()) {
-      return idcsLogin(request);
-    }
-    return login(request, idcsClient)
-        .addingToSession(request, REDIRECT_TO_SESSION_KEY, redirectTo.get());
+  public Result adminLogin(Http.Request request) {
+    return login(request, adminClient);
   }
 
-  public Result loginRadiusLogin(Http.Request request) {
-    return login(request, loginRadiusClient);
-  }
-
-  public Result loginRadiusLoginWithRedirect(Http.Request request, Optional<String> redirectTo) {
+  public Result applicantLogin(Http.Request request, Optional<String> redirectTo) {
     if (redirectTo.isEmpty()) {
-      return loginRadiusLogin(request);
+      return login(request, applicantClient);
     }
-    return login(request, loginRadiusClient)
+    return login(request, applicantClient)
         .addingToSession(request, REDIRECT_TO_SESSION_KEY, redirectTo.get());
   }
 
   public Result register(Http.Request request) {
-    String registerUrl = null;
+    String idpProvider;
     try {
-      registerUrl = config.getString("idcs.register_uri");
+      idpProvider = config.getString("auth.applicant_idp");
     } catch (ConfigException.Missing e) {
-      // leave it as null / empty.
+      // Default to IDCS
+      idpProvider = AuthIdentityProviderName.IDCS_APPLICANT.toString();
     }
-    if (Strings.isNullOrEmpty(registerUrl)) {
-      return badRequest("Registration is not enabled.");
+    // This register behavior is specific to IDCS. Because this is only being called when we know
```

Comment:
I'm not super clear on why we only need it for IDCS, I just know that we don't for LoginRadius. I think it's just how different IDPs work

Reply:
Suggestion:

restructure/rephrase this similar to the following to return early in the common case, and draw out the exception more.

boolean isIDCS = idp.equals(AuthIdentityProviderName.IDCS_APPLICANT.toString());

if(!isIDCS) {
  return login(request, applicantClient);
}

// IDCS requires a special registration flow.
.......


-----------
This might also indicate that the IDCS code blog should be a method that is called here to compartmentalize things.

Reply:
I restructured and tagged you in the comment, let me know what you think